liquidity_pool_router_with_payload/src/lib.rs

```diff
@@ -0,0 +1,242 @@
+
+        let reserves = pool_client.get_rsrvs();
+        let amounts = get_deposit_amounts(
+            deposit_a.desired,
+            deposit_a.min,
+            deposit_b.desired,
+            deposit_b.min,
+            reserves,
+        );
+
+        let mut sigs_a = Map::new(&e);
+        sigs_a.set(
+            symbol!("sig"),
+            sigs.get_unchecked(symbol!("xfer_a")).unwrap(),
+        );
+        let mut sigs_b = Map::new(&e);
+        sigs_b.set(
+            symbol!("sig"),
+            sigs.get_unchecked(symbol!("xfer_b")).unwrap(),
+        );
+
+        token_contract::Client::new(&e, deposit_a.token).xfer(
+            &to,
+            &deposit_a.nonce,
+            &Identifier::Contract(e.get_current_contract()),
+            &amounts.0,
+            &sigs_a,
+        );
+        token_contract::Client::new(&e, deposit_b.token).xfer(
+            &to,
+            &deposit_b.nonce,
+            &Identifier::Contract(e.get_current_contract()),
+            &amounts.1,
+            &sigs_b,
+        );
+
+        pool_client.deposit(&to);
+    }
+}
+
+pub trait PayloadTrait {
+    fn has_sig(e: Env, function: Symbol) -> bool;
+
+    fn payload(
+        e: Env,
+        function: Symbol,
+        args: Vec<RawVal>,
+        callstack: Vec<(BytesN<32>, Symbol)>,
+    ) -> Map<Symbol, (Identifier, SaltedSignaturePayload)>;
+}
+
+pub struct LiquidityPoolRouterPayload;
+
+#[contractimpl]
+impl PayloadTrait for LiquidityPoolRouterPayload {
+    fn has_sig(_e: Env, function: Symbol) -> bool {
+        const DEPOSIT_RAW: u64 = symbol!("deposit").to_raw().get_payload();
+        match function.to_raw().get_payload() {
+            DEPOSIT_RAW => true,
+            _ => false,
+        }
+    }
+
+    fn payload(
+        e: Env,
+        function: Symbol,
+        args: Vec<RawVal>,
+        callstack: Vec<(BytesN<32>, Symbol)>,
+    ) -> Map<Symbol, (Identifier, SaltedSignaturePayload)> {
+        const DEPOSIT_RAW: u64 = symbol!("deposit").to_raw().get_payload();
+        match function.to_raw().get_payload() {
+            DEPOSIT_RAW => {
+                let to: Identifier = args.get_unchecked(0).unwrap().try_into_val(&e).unwrap();
+                let deposit_a: Deposit = args.get_unchecked(1).unwrap().try_into_val(&e).unwrap();
+                let deposit_b: Deposit = args.get_unchecked(2).unwrap().try_into_val(&e).unwrap();
+
+                let pool_id = get_pool_id(&e, &pool_salt(&e, &deposit_a.token, &deposit_b.token));
+                let reserves = pool_contract::Client::new(&e, &pool_id).get_rsrvs();
+                let amounts = get_deposit_amounts(
+                    deposit_a.desired,
+                    deposit_a.min,
+                    deposit_b.desired,
+                    deposit_b.min,
+                    reserves,
+                );
+
+                let mut callstack_a = callstack.clone();
+                callstack_a.push_back((deposit_a.token.clone(), symbol!("xfer")));
```

Comment:
Should call stack a contain this contract (i.e. liquidity pool)?

Reply:
Great question. I think it could go either way. Contracts would probably be mildly cheaper if we defined "call stack" to mean everything before this function--this means the simple case of passing a signature to the first contract incurs no call stack cost since it would be empty. But the whole system would probably be slightly less surprising if we defined "call stack" to mean everything including this function. What do you think?